feat(ui): use builtin completion popupmenu with ext_cmdline

luukvbaal opened the pull request:
Problem:  UIs implementing ext_cmdline/message must also implement
          ext_popupmenu in order to get cmdline completion with
          wildoptions+=pum.
Solution: Allow marking a window as the ext_cmdline window through
          nvim_open_win(), including prompt offset. Anchor the cmdline-
          completion popupmenu to this window.

luukvbaal commented on src/nvim/api/win_config.c:
```diff
@@ -212,6 +212,10 @@
+///   - cmdline: If true, indicate that this window is used for ext_cmdline.
+///     Nvim will anchor the 'wildoptions' popupmenu to this window.
+///   - cmdline_offset: Offset in units of "screen cell width" indicating how much longer
+///     (negative for shorter) the displayed prompt is from the prompt emitted by "cmdline_show".
```
Could keep these private for use with the default ext_cmdline (`_cmdline/_cmdline_offset`) but after testing this seems useful enough to expose.

justinmk replied:
If we expose these, they should be options (for reference, the existing 'previewwindow' option makes a similar kind of decision). Something like a new, global 'popupwindow' option (or 'popupopt', 'pmenuopt', ...).

Setting a global option from an unrelated "create"/"open" API is strange.

If it helps unblock this PR, for now let's just not expose these.

luukvbaal replied:
Still awaiting @bfredl's review, who originally suggested adding this to `nvim_open_win` :) From matrix:

>> I was also wondering if some API to "set" the cmdline window would be better for some internal work Nvim can do. But perhaps documenting that the implementation should move the cursor there to this end is sufficient?

> I think this can be done, but being future-proof for more "roles" being added later. perhaps a field in nvim_win_set_config/win_open

justinmk replied:
`nvim_win_set_config` would fit better, since it is essentially window-local properties, which is very similar to window-local options. Though a global option seems to make more sense than a window-local option, since a global option is easier to "query".

luukvbaal replied:
`nvim_win_set_config` and `nvim_open_win` share the same `opts` dictionary, so adding it to one implies the other. I'm not sure how a global option would work, do you mean that option would hold the window handle?

justinmk replied:
> do you mean that option would hold the window handle?

Yes. That seems natural, no?

luukvbaal replied:
Yes if it's a global option it would have to be set/unset whenever the window is deleted (which the user could do, at which point the handler would have to create a new window). So unless we disallow deleting a window marked as such, attaching the flag to the window itself seems to make more sense to me.

luukvbaal replied:
There is something to say for symmetry with `previewwindow` though, so should probably promote it to an actual window option if @bfredl agrees. This does bring into question the entire `WinConfig` struct.

justinmk replied:
> This does bring into question the entire `WinConfig` struct.

Let's save that for https://github.com/neovim/neovim/issues/30453 , out of scope here.

bfredl replied:
I'd say keep it a `_` field in winconfig for now.

przepompownia commented on test/functional/lua/ui_event_spec.lua:
```diff
@@ -306,6 +306,92 @@ describe('vim.ui_attach', function()
+          vim.api.nvim_win_set_cursor(_G.win, {1, #content[1][2]})
+          vim.api.nvim__redraw( { win = _G.win, cursor = true, flush = true })
+        end
+        return true
```
`:help vim.ui_attach()` doesn't seem to mention the meaning of the return value.

luukvbaal replied:
Yeah, it makes Nvim not send the event to remote UIs. I'm not sure what the reason behind it is. Here is a related issue: #21150.

przepompownia replied:
Indeed (if I understand correctly): the test case uses remote UI and cannot pass without it. By comparison, at TUI it doesn't matter.

luukvbaal replied:
This test does not test the remote UI state. In this case specifically returning `true` is used to hide the ext_ state from the screen:expect output, the screen class is the remote UI.

justinmk commented on src/nvim/ex_getln.c:
```diff
@@ -2942,10 +2942,9 @@ static int cmd_startcol(void)
   int col = cmd_startcol();
   if (KeyTyped) {
-    m = Columns * Rows;
+    m = cmdline_win ? cmdline_win->w_width_inner * cmdline_win->w_height_inner : Columns * Rows;
```
Would it make sense to *always* define `cmdline_win`? The default impl would anchor its dimensions to `Columns * Rows`, but that impl would be contained in the default `cmdline_win` handler, so code never needs to check if `cmdline_win` exists:

```suggestion
    m = cmdline_win->w_width_inner * cmdline_win->w_height_inner;
```

I don't know if this is a net gain, so this is suggestion is also a question :)

luukvbaal replied:
Don't think so; it would have to be a grid not a window (only the default grid relates to `Columns/Rows`). But it's not straightforward to get the inner dimensions from the grid, and we can't query the cmdline row/column from the default grid, so would need ternaries anyways.

justinmk commented on src/nvim/api/win_config.c:
```diff
@@ -212,6 +212,10 @@
 ///   - vertical: Split vertically |:vertical|.
 ///   - split: Split direction: "left", "right", "above", "below".
+///   - cmdline: If true, indicate that this window is used for ext_cmdline.
+///     Nvim will anchor the 'wildoptions' popupmenu to this window.
```
Not clear from this description what the window shows. Is it only for the popup menu? `ext_cmdline` presents the entire cmdline, not only the popup menu.

luukvbaal replied:
What the window contains is entirely up to the cmdline implementation. This just allows the implementation to "mark" the window that is used to display the cmdline as such, to allow Nvim to do some work with it internally.

For now that just means anchoring the completion popupmenu, not sure if anything else could be done. Originally also thought this could be used to redraw and move the cursor to it, but since we can do that with `nvim__redraw` this is no longer necessary.

justinmk commented on src/nvim/buffer_defs.h:
```diff
@@ -958,6 +958,11 @@ typedef struct {
+/// longer (negative for shorter) the displayed prompt is from the prompt
+/// emitted by "cmdline_show". Used to position the |cmdline-completion| popup
+/// window with ext_cmdline.
+  int _cmdline_offset;
```
comment indent looks wrong

justinmk commented on test/functional/lua/ui_event_spec.lua:
```diff
@@ -306,6 +306,90 @@ describe('vim.ui_attach', function()
+        style = 'minimal',
+        focusable = false,
+        zindex = 300,
+        _cmdline_offset = -1, -- not inserting "firstc" below
```
do we need a test for some other offset? maybe using `nvim_win_set_config`?

justinmk commented on src/nvim/buffer_defs.h:
```diff
@@ -958,6 +958,11 @@ typedef struct {
   bool hide;
+/// cmdline_offset: Offset in units of "screen cell width" indicating how much
+/// longer (negative for shorter) the displayed prompt is from the prompt
+/// emitted by "cmdline_show". Used to position the |cmdline-completion| popup
+/// window with ext_cmdline.
```
"Setting this activates ..."

bfredl commented on src/nvim/buffer_defs.h:
```diff
@@ -958,6 +958,11 @@ typedef struct {
   bool noautocmd;
   bool fixed;
   bool hide;
+/// cmdline_offset: Offset in units of "screen cell width" indicating how much
```
no need for the quotes, cell width is a well known concept used everywhere already.

luukvbaal replied:
Sure, but IIRC I copied this verbatim from nvim_open_win.

bfredl commented on test/functional/ui/cmdline_spec.lua:
```diff
@@ -623,8 +623,7 @@ local function test_cmdline(linegrid)
 
-    command('set wildmode=full')
-    command('set wildmenu')
+    command('set wildmode=full wildmenu wildoptions=pum')
```
Is this just a clean up of the tests to make them more explicit, or is this necessary due to a change of behavior?

luukvbaal replied:
Includes a behavior change. In this PR ext_popupmenu doesn't (I thought it shouldn't) receive events unless the user sets wildoptions=pum. ext_wildmenu still does regardless of the option.

bfredl replied:
ext_popupmenu also includes wildmenu (regardless of options):

> - `ext_popupmenu`	Externalize |popupmenu-completion| and
>			'wildmenu'. |ui-popupmenu|

Adding a separate `ext_wildmenu` format  was a mistake and this option is now deprecated.

luukvbaal replied:
Yes but if the user indicates they don't want a popupmenu for the cmdline through wildoptions, why would we still emit the event?

justinmk replied:
Behavior changes LGTM, ok w/ @bfredl ?

bfredl replied:
No, this should not be changed. what `ext_popupmenu` actually means is forward **the data** related to multiple-choice completion. `ext_popupmenu` does emphatically not refer to the one particular visual style of a menu (which TUI vim has called "the popupmenu" in big singularis) Thus whether it is used is irrelevant of `wildoptions+=pum` is set or not, which is a _newer option_ meant to control legacy TUI only.

justinmk commented on src/nvim/api/win_config.c:
```diff
@@ -295,6 +295,10 @@ Window nvim_open_win(Buffer buffer, Boolean enter, Dict(win_config) *config, Err
     goto cleanup;
   }
 
+  if (fconfig._cmdline_offset < INT_MAX) {
```
In the `nvim_open_win` we should mention the flag and mention "EXPERIMENTAL". We don't have a better way of documenting experiemental stuff currently. But not documenting it at all means it could be forgotten and/or lead to confusion in the future.

przepompownia commented:
I started using this and haven't noticed any special issues for now.

przepompownia commented:
Resolving the conflict would make local merging easier.

luukvbaal committed:
feat(ui): use builtin completion popupmenu with ext_cmdline

Problem:  UIs implementing ext_cmdline/message must also implement
          ext_popupmenu in order to get cmdline completion with
          wildoptions+=pum.
Solution: Allow marking a window as the ext_cmdline window through
          nvim_open_win(), including prompt offset. Anchor the cmdline-
          completion popupmenu to this window.

przepompownia commented:
@luukvbaal let me not guess how to resolve (at local merge) the conflict with recent multigrid info changes :grin:
